File: research/experiments/sweep/runs/opt-small-kotlin-sonnet45/src/main/kotlin/optimization/VecOps.kt
package optimization

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Pure vector arithmetic for n-dimensional optimization.
 * All operations return new arrays and never mutate inputs.
 */
object VecOps {
    /**
     * Dot product of two vectors.
     */
    fun dot(a: DoubleArray, b: DoubleArray): Double {
        require(a.size == b.size) { "Vectors must have same length" }
        var sum = 0.0
        for (i in a.indices) {
            sum += a[i] * b[i]
        }
        return sum
    }

    /**
     * Euclidean (L2) norm.
     */
    fun norm(v: DoubleArray): Double {
        return sqrt(dot(v, v))
    }

    /**
     * Infinity norm (max absolute value).
     */
    fun normInf(v: DoubleArray): Double {
        var max = 0.0
        for (x in v) {
            val absX = abs(x)
            if (absX > max) {
                max = absX
            }
        }
        return max
    }

    /**
     * Scalar multiplication.
     */
    fun scale(v: DoubleArray, s: Double): DoubleArray {
        val result = DoubleArray(v.size)
        for (i in v.indices) {
            result[i] = v[i] * s
        }
        return result
    }

    /**
     * Element-wise addition.
     */
    fun add(a: DoubleArray, b: DoubleArray): DoubleArray {
        require(a.size == b.size) { "Vectors must have same length" }
        val result = DoubleArray(a.size)
        for (i in a.indices) {
            result[i] = a[i] + b[i]
        }
        return result
    }

    /**
     * Element-wise subtraction.
     */
    fun sub(a: DoubleArray, b: DoubleArray): DoubleArray {
        require(a.size == b.size) { "Vectors must have same length" }
        val result = DoubleArray(a.size)
        for (i in a.indices) {
            result[i] = a[i] - b[i]
        }
        return result
    }

    /**
     * Element-wise negation (scale(v, -1)).
     */
    fun negate(v: DoubleArray): DoubleArray {
        return scale(v, -1.0)
    }

    /**
     * Deep copy.
     */
    fun clone(v: DoubleArray): DoubleArray {
        return v.copyOf()
    }

    /**
     * Vector of n zeros.
     */
    fun zeros(n: Int): DoubleArray {
        return DoubleArray(n)
    }

    /**
     * Fused operation: a + s*b (avoids intermediate allocation).
     */
    fun addScaled(a: DoubleArray, b: DoubleArray, s: Double): DoubleArray {
        require(a.size == b.size) { "Vectors must have same length" }
        val result = DoubleArray(a.size)
        for (i in a.indices) {
            result[i] = a[i] + s * b[i]
        }
        return result
    }
}

File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/VecOps.cs
namespace Optimization;

/// <summary>
/// Pure vector arithmetic for n-dimensional optimization.
/// All operations return new arrays and never mutate inputs.
/// </summary>
public static class VecOps
{
    public static double Dot(double[] a, double[] b)
    {
        double sum = 0;
        for (int i = 0; i < a.Length; i++) sum += a[i] * b[i];
        return sum;
    }

    public static double Norm(double[] v)
    {
        return Math.Sqrt(Dot(v, v));
    }

    public static double NormInf(double[] v)
    {
        double max = 0;
        for (int i = 0; i < v.Length; i++)
        {
            double abs = Math.Abs(v[i]);
            if (abs > max) max = abs;
        }
        return max;
    }

    public static double[] Scale(double[] v, double s)
    {
        double[] result = new double[v.Length];
        for (int i = 0; i < v.Length; i++) result[i] = v[i] * s;
        return result;
    }

    public static double[] Add(double[] a, double[] b)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++) result[i] = a[i] + b[i];
        return result;
    }

    public static double[] Sub(double[] a, double[] b)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++) result[i] = a[i] - b[i];
        return result;
    }

    public static double[] Negate(double[] v)
    {
        return Scale(v, -1);
    }

    public static double[] Clone(double[] v)
    {
        return (double[])v.Clone();
    }

    public static double[] Zeros(int n)
    {
        return new double[n];
    }

    public static double[] AddScaled(double[] a, double[] b, double s)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++) result[i] = a[i] + s * b[i];
        return result;
    }
}
